relayer/src/keyring.rs

```diff
@@ -343,15 +345,33 @@ impl KeyRing {
     }
 
     /// Sign a message
-    pub fn sign_msg(&self, key_name: &str, msg: Vec<u8>) -> Result<Vec<u8>, Error> {
+    pub fn sign_msg(
+        &self,
+        key_name: &str,
+        msg: Vec<u8>,
+        address_type: &AddressType,
+    ) -> Result<Vec<u8>, Error> {
         let key = self.get_key(key_name)?;
 
         let private_key_bytes = key.private_key.private_key.to_bytes();
-        let signing_key =
-            SigningKey::from_bytes(private_key_bytes.as_slice()).map_err(Error::invalid_key)?;
-
-        let signature: Signature = signing_key.sign(&msg);
-        Ok(signature.as_ref().to_vec())
+        match address_type {
+            AddressType::Cosmos => {
+                let signing_key = SigningKey::from_bytes(private_key_bytes.as_slice())
+                    .map_err(Error::invalid_key)?;
+                let signature: Signature = signing_key.sign(&msg);
+                Ok(signature.as_ref().to_vec())
+            }
+            AddressType::Ethermint { .. } => {
+                let hash = keccak256_hash(msg.as_slice());
```

Comment:
https://github.com/tharsis/ethermint/blob/main/crypto/ethsecp256k1/ethsecp256k1.go#L110
ethermint only do the keccak hash if the msg length is not 32, but in `VerifySignature` it didn't check the msg length though, could be a bug in ethermint side.

Reply:
when I look at go-ethereum, `sign...` will always hash the input (tx, data...) and then call `signHash` to produce the signature

Reply:
I never noticed the keccak hash was conditional. Our codebase that has been forked from ethermint at some time didn't have this.

Moreover, this is the first time I ever see that flow in crypto libs xD

Reply:
Some interesting forgery attacks on ECDSA when the message isn't hashed before signing (i.e. when message length is 32 bytes) -> https://crypto.stackexchange.com/questions/44862/ecdsa-signature-without-hashing-or-with-offloaded-hash